rero_ils/modules/contributions/sync.py

```diff
@@ -0,0 +1,546 @@
+from rero_ils.modules.utils import get_timestamp, set_timestamp
+
+
+class SyncAgent(object):
+    """Agent MEF synchronization."""
+
+    def __init__(self, dry_run=False, verbose=False, log_dir=None,
+                 from_last_date=False):
+        """Constructor.
+
+        :param dry_run: bool - if true the data are not modified
+        :param verbose: bool or integer - verbose level
+        :param log_dir: string - path to put the logs
+        :param from_last_date: boolean - if True try to consider agent modified
+            after the last run date time
+        """
+        self.dry_run = dry_run
+        self.verbose = verbose
+        self.from_date = None
+        self.start_timestamp = None
+        self._init_logger(verbose, log_dir)
+        if from_last_date:
+            self._get_last_date()
+
+    def _get_last_date(self):
+        """Get the date of the last execution of the synchronization."""
+        self.from_date = None
+        data = get_timestamp('sync_agents')
+        if data and data.get('start_timestamp'):
+            self.from_date = data.get('start_timestamp')
+
+    def _init_logger(self, verbose, log_dir):
+        """Initialize the module logger."""
+        # default value
+        if not log_dir:
+            log_dir = current_app.config.get(
+                'RERO_ILS_MEF_SYNC_LOG_DIR',
+                os.path.join(current_app.instance_path, 'logs')
+            )
+        # create the log directory if does not exists
+        if not os.path.exists(log_dir):
+            os.mkdir(log_dir)
+        verbose_level = ['ERROR', 'INFO', 'DEBUG']
+        logging_config = {
+            'version': 1,
+            'disable_existing_loggers': False,
+            'formatters': {
+                'standard': {
+                    'format': '%(asctime)s [%(levelname)s] :: %(message)s'
+                }
+            },
+            'handlers': {
+                'console': {
+                    'class': 'logging.StreamHandler',
+                    'formatter': 'standard',
+                    'level': verbose_level[min(verbose, 2)]
+                },
+                'file': {
+                    'class': 'logging.handlers.TimedRotatingFileHandler',
+                    'filename': os.path.join(log_dir, 'sync_mef.log'),
+                    'when': 'D',
+                    'interval': 7,
+                    'backupCount': 10,
+                    'formatter': 'standard'
+                }
+            },
+            'loggers': {
+                # for the current module
+                self.__module__: {
+                    'handlers': ['console', 'file'],
+                    'level': 'INFO',
+                    'propagate': False
+                }
+            }
+        }
+        logging.config.dictConfig(logging_config)
+        self.logger = logging.getLogger(__name__)
+        self.log_dir = log_dir
+
+    def agent_are_different(self, agent1, agent2):
```

Comment:
Multiple remarks :
  * is it a good candidate for `__cmp__` special method implementation ?
  * if you choose to use 2 agents parameters, this method should be a `@classmethod` (move logging ?)
  * if you would keep self (for logging), use only one `agent` params.
  
  As you wrote it, it's just like you have 1 apple to compare 2 other apples

Reply:
1. I do not think that it is a good candidate for `__cmp__` unless to create a new classe representing an agent which can be sometime, a Contribution object or a dict. And here it is a very specific comparison as we remove some fields such as `organisation.

Here self is not an agent by a Sync object processing this like an human that compare 2 apples.